### DISCUSSION TITLE
what are ports?

### DISCUSSION THREAD

**Discussion by tjdphd** [thliebig/openEMS-Project#126] [General]
forgive me for being a bit slow but I don't understand what the manual says about ports. I.e., what they are and what they're for. The manual says they "are macro function". First of all, do you mean "macro functions". I.e. should the word "function" be pluralized there? If not what does "macro function" mean?  Secondly, in what sense are they "macro?" Thirdly you say they "combine e.g. current and voltage probes...". Combine them how? combine them where? Why do they need to be "combined?" what are they combined into? This is completely opaque to me. Are they meant to specify the placement of probes or other kinds of devices?  What happens if you don't have any ports? 

thank you!

**Comment by biergaizi** [comment]
Do you have any experience on measuring microwave circuits using a Vector Network Analyzer (VNA), or doing S-parameter simulations in a lumped RF circuit simulator like ADS or Qucs? The "port" in an openEMS simulation basically serves the same propose like the port on a VNA, with three functions:

1. Inject a signal into the circuit, with an output resistance (usually matched to the system impedance, like 50 Ω)
2. Receive the signal coming out from the other side of the circuit, with an input resistance (usually matched to the system impedance, like 50 Ω).
5. Record the voltage and current waves  - recorded waveforms can then be used to calculate circuit parameters of a microwave device, such as S11 (complex reflection coefficient) and S21 (complex transmission coefficient). After the full n-port S-parameters are obtained, the entire device can then be modeled as a blackbox (assume linearity) and used in a standard RF lumped circuit simulator.

A port in openEMS is basically an idealized version of the same thing. 

> A lumped port is a 1D, 2D or 3D box containing a feeding resistance (e.g. 50 Ohms) and a current and voltage probe measuring the current through and the voltage along this resistance. As any port it can be active/include an [excitation](https://wiki.openems.de/index.php/Excitation.html) or not.

**Reply by tjdphd** [reply]
Thanks very much for responding! Regrettably, no, I have no such experience. I've had graduate level Electromagnetism (I.e. Jackson) but no experience, real or virtual with related devices.  I guess I'm trying to get an intuitive sense of what a "port" is in relation to the physical device/object that one is modeling. 

while I sincerely appreciate what you've written, it only tells me what a port is "for" without providing any intuition about what a port "is," or how I should specify a port for a given application. It's clear enough from what I've been able to find in the documentation that there are different kinds of ports for different circumstances. Okay, but why? do ports have shapes and sizes? Are they meant to represent physical objects, or just portions of the simulation domain where something special happens? 

Let's take the example of the ports that are specified for the Rectangular waveguide example in the tutorial. (I'm looking at the python interface here).  Here's what one of the specifications looks like:

start=[0, 0, 10*mesh_res];
stop =[a, b, 15*mesh_res];
mesh.AddLine('z', [start[2], stop[2]])
ports.append(FDTD.AddRectWaveGuidePort( 0, start, stop, 'z', a*unit, b*unit, TE_mode, 1))

Where a and b are just the waveguide width and height respectively and mesh_res is the targeted mesh resolution set to be 1/30 of wavelength corresponding to the center of a frequency band. 

 So what's going on here? What is "start" and what is "stop"?  What do the numbers in their specification mean? Are they Cartesian coordinates?  Why is mesh_res used for specifying the third list elements in start and stop? Why are the specifications different for start and stop? Does this specify a thickness and are the thicknesses different? 

What is AddLine? Is that part of the port specification? What is the meaning of the '0' in the first argument of AddRectWaveGuidePort refer to?  What does the 'z' mean? (Really it seems to me that an explanation for all of the arguments to this function is called for.) 

I guess it's the documentarian's expectation that I'm supposed to be able to just infer what's going on here, but, sorry, I just can't.  

Even after all this is specified, what exactly have I specified? You say it's for injecting or receiving a signal (among other things). Earlier in the example there is a Gaussian excitation specified, So, does this mean that this Gaussian signal is specified to occur in the region of the simulation domain corresponding to the location specified for the port? How does openEMS know which of the two ports is originating the signal and which is receiving? 

I know I'm asking really elementary questions, and I've looked all over the internet without success for an explanation that makes sense to me, but everything I find already seems to assume knowledge and/or intuition that I don't have yet. I even found an entire textbook on the subject of the FDTD method, but it says not one word about "ports."  I'd really appreciate a "for dummies" explanation if you've got the time, because like it or not I need to be able to use openEMS for work.  Thanks again for any help you can offer!

**Reply by LubomirJagos** [reply]
Hi, reading @biergaizi answer I was laughing for while because I was expecting your next comment :D :D his answer is good but in my opinion not too practical (I appreciate it, that was just joke what I wrote in previous sentence). So my answer for dummies is next.
    1. in FDTD algorithm everything is happening around model which is created by nodes in rectangular grid, every computation is in these nodes
    2. port is just area, as you wrote that example with rectangular waveguide port it means that that port is created with all nodes which are included in box with corners at start, stop (in openEMS there would be just plane generated as input of that waveguide but that something else), so port is defined in some space in model and it means that all calculation is done using these nodes
    3. now when port is active, simplest case is it's excited by sine signal it means that EVERY node has in time its Efield defined by equation amplitude*sin(2*pi*f0), if gaussian excitation is used then every node at time has Efield with value gauss_function(2*pi*f0), and so this is like when you put somewhere output from signal generator, at the input is electric signal from generator
    - also when port has specified resistance then each node included in port have this resistance which is used in calculation
    - so active port is like hardcoded condition for port at each time moment
    4. now when port is inactive, it means it's used as probe and it means whe you are reading voltage volume integral is calculated over nodes and E and H field are integrated and result is scalar number like 5.3V, 4.7V or whatever, same for current and also it's resistance is taking into consideration when these values are calculated (FDTD itself using E and H fields is internaly using V/m and A/m due they are in space)
    
So ports are just groups of nodes in space when port is active fields values are forced into model by specified function, when they are inactive values can be extracted from them using volume integral.

**Reply by tjdphd** [reply]
Thanks @LubomirJagos! Yes! This is the sort of answer I'm looking for. So, a few more questions then, if you don't mind:

1.) Does this mean that in the example I gave 10*mesh_res is the z coordinate of the port corner "start" and that 15*mesh_res is the z coordinate of the port corner "stop" and that therefore the port has a thickness of 5*mesh_res? 

2.) Is one of the arguments in the method AddRectWaveGuidePort meant to specify whether the port is active or not? If so, which one? 

3.) How do you know if a port is "lumped" or not and what is an example of a port that is not "lumped." What intuition should I associate with this use of the word "lumped?"  Also, would it be correct to say that ports are either lumped or they're not lumped? 

4.) what mischief would I cause if I were to replace 'z' with say 'x' in the call to AddRectWaveGuidePort? 

5.) also I'm still unclear on what that addLine call is about and whether it's important or not. 

6.) Would it be fair to say that the ports defined in Rect_Waveguide example serve to specify those locations corresponding to the location of the physical waveguide being modeled itself? Or to put it another way, does the specification of the ports given in this example constitute the actual model of the waveguide? 

Anyway, thanks again for your answer.

**Reply by LubomirJagos** [reply]
ok, just to be said @biergaizi is master at paralelizing openEMS code and he is doing job I can just dream, your questions are ok, everybody is asking these questions, I myself was looking and experimenting with scripts to make myself sure how engine is running and what is happening inside model.

Now little theory needed for understanding E and H fields in FDTD, at each node E field consist of Ex, Ey, Ez H field consists Hx, Hy, Hz and Ex, Ey, Ez, Hx, Hy, Hz are all complex numbers. So E and H are complex vectors and each x,y,z parts are also complex vectors (electromagnetism is bitch) because electric and magnetic field in space has amplitude and phase :)

Here are answers:
1. yes, in z ax it contains 5 lines and in x, y axes lines between 0..a and 0..b, grid is specified in mesh structure there are literally mesh.x = [0, 0.2, 0.3 or whatever], mesh.y = [0, 0.33, 0.45 or whatever]

2. yes there is parameter in python (you used python there is also matlab/octave interface) excite which can be 0 for inactive port or some amplitude number like 4.3 what means 4.3V/m for E field amplitude

3. internally inside FDTD engine everything about nodes is just Efield and Hfield values in nodes and their other parameters (permeability, permitivity and resistance) just influence values of E and H, port kind just specify for active port how values are assigned to nodes inside port so:
lumped port - excitation can be defined JUST IN ONE DIRECTION means x,y, or z and each port node has assigned at each time same value specified by sin(), gauss() or impulse() or custom function, this is most easiest port as everybody everytime imagine ports
waveguide port - on input plane (rectangle or circle based on waveguide shape) values of E and H are weighted (there is under hood in openEMS field weighting function for E and H fields for port) and there is equation for each mode, mode is solution of differential equation which can exists and traverse through waveguide it has shapes like simple O or double O or butterfly or whatever funny stuff what learned at university (they spend loooot of time trying to explained to us :D) even electrons in atoms have these shapes due they are waves if they want to be :D that input equation is not hard it's just somehow grading E field like sinc() or something you can find that equation on internet for basic modes like TE10, TE11... look for code inside python interface function but I think that matlab interface is much more readible for this

4. if you mischief port direction it stops running and you will got invalid simulation, propagation direction is direction in which waveguide is going (you can try to put into simulation x, y, z and it will give you right results just for one :D best is to try it)

5. it's adding gridlines to z coord to have intersection with rest gridlines in x,y to have nodes at edge of waveguide, in your code snippet it's not important, for whole simulation it's important, for this discussion that line is misleading and unecessary


6. yes, in this example it is part of waveguide, this port is introducing into model E,H field from waveguide and it should be located at the input hole of waveguide model modelled using PEC or metal or some structure, it's like output hole from one waveguide going into your model
    
Like jumping into openEMS is kind hard, too hard, it tooks me first time like 3 months to write running script itself :D :D :D and another months to fully understand after lot of experimenting how to use it properly, so I created GUI addon which makes using openEMS similar to HFSS or CST, maybe you can try it and experimenting with simulations give you answers or you can ask here.

OpenEMS Simlation Creator for FreeCAD addon: https://github.com/LubomirJagos/FreeCAD-OpenEMS-Export

in released version properly running generator for code is just matlab/octave, if you want to use python I almost finished it (now it's just missing some ports and cylindrical coords but you should start with cartesian) you have to switch it for develop branch, there is in left top corner combobx at github, here is video how to install it, I should make another how to model some models, I think for beginning and most wanted is to try model planar patch antenna or filter with lumped port, after it everything will be more obvious:
https://www.youtube.com/watch?v=OR-79dCMdyo

**Reply by tjdphd** [reply]
Thanks again for all your help! I'll have to spend some time ingesting all this, and maybe I'll be back with more questions. I'll definitely checkout  your simulation creator. Thanks for that too!  I do very much need to use the python interface but I'm worried about whether I'll be able to because it doesn't appear to include the option for a custom excitation, but we'll see. (I started another thread about that.) Anyway, thank you very much for your help!

**Reply by LubomirJagos** [reply]
:D :D yes I was answering that one too :D haven't tried custom excitation in python interface but today I finished most of python code generator and I want to add this one there too, yes there is no function for it now, but in python everything (all variables and class members functions) is public, so you can access underlying CSXCAD methods and properties so it's just juggling with underlying methods

so final answer is yes it's possible to do custom excitation in python but it's hidden, not to deep just need like day to play with it and spell will be broken :D

**Comment by tjdphd** [comment]
Hello again @LubomirJagos  I've been trying to follow along with your YouTube video, but even when I have a new file open I get an error when I try to run the macro. The exact error I get is as follows:

16:53:18  Traceback (most recent call last):
  File "C:/Users/tdennis/repos/FreeCAD-OpenEMS-Export-main/FreeCAD-OpenEMS-Export.FCMacro", line 25, in <module>
    FOE_spec.loader.exec_module(FOE_mod)
  File "<frozen importlib._bootstrap_external>", line 839, in exec_module
  File "<frozen importlib._bootstrap_external>", line 975, in get_code
  File "<frozen importlib._bootstrap_external>", line 1032, in get_data
<class 'FileNotFoundError'>: [Errno 2] No such file or directory: 'C:\\Users\\tdennis\\repos\\FreeCAD-OpenEMS-Export-main\\FreeCAD-OpenEMS-Export/ExportOpenEMSDialog.py'

It seems to think that the python script ExportOpenEMSDialog.py should be a subdirectory of the FreeCAD-OpenEMS-Export-main directory called "FreeCAD-OpenEMS-Export" but when I extracted the plugin zip file I don't find such a subdirectory. Instead the python script is in the same directory as the macro as you can see below:

 ls FreeCAD-OpenEMS-Export-main
__init__.py     documentation_img/  ExportOpenEMSDialog.py           img/     README.md  ui/
documentation/  experiments/        FreeCAD-OpenEMS-Export.FCMacro*  LICENSE  test/      utilsOpenEMS/

Any idea what I might be doing wrong?

Thanks!

**Reply by LubomirJagos42** [reply]
Hi, as you write that file FreeCAD-OpenEMS-Export.FMacro is just python file launcher which internally opens ExportOpenEMSDialog.py, so now you have 2. options:
    1. in freecad opens in macro ExportOpenEMSDialog.py and it will run
    2. that file FreeCAD-OpenEMS-Export.FMacro copy folder up, and rename folde with addon as FreeCAD-OpenEMS-Export, means remove -main. The idea why there is launcher file is that as this addon is programmed as macro it should be in folder with other freecad macros and therefore original way is that you download addon unpack it into folder but you need launcher which you copy from folder to above folder to have it alongside with the others macros and therefore still have access in freecad to others macro and dont need to switch folder every time when want to use it

**Reply by tjdphd** [reply]
I can't really understand your answer very well, but I was able to get the Simulation Creator gui to open by running the python script ExportOpenEMSDialog.py directly instead of trying to run the macro.  That said, I don't really know what to do next. It would be nice if you had a video showing the actual creation of a simulation file.

**Reply by LubomirJagos42** [reply]
sure, I would be making one since there is some MakerFair in Brno in Czech republic so I want to add them but there were other things to do, this is like structure similar to other commercial tools, it's not that hard if you already worked with other simulations, this is kind similar to AnsoftHFSS
